### PR TITLE
[Core] Card component

### PR DESCRIPTION
**Pull request by JacekJagiello**
Hey, after @giladgray introduced `Callout` and `FormGroup` components, I think it's time to add one more little thing, the `Card` component.

#### Changes proposed in this pull request:
Added `Card` component with extra loading feature, tests, and docs

#### Reviewers should focus on:
* Should interactive mode be disabled, while loading? [For now, both state can be used simultaneously]
* Should onClick only be called, when card is in interactive mode? [For now, it's always called; it doesn't depends on interactive mode]
* ~While loading, should `tabindex="-1"` be added for every child of the card component? [For now, it's added for every single node under card component]~

Live preview:
https://23-95210685-gh.circle-artifacts.com/0/home/ubuntu/blueprint/packages/site-docs/dist/index.html#core/components/card.javascript-api

#### Screenshot
card-component

#### Checklist

- [x] [Enable CircleCI for your fork](https://circleci.com/add-projects)
- [x] Include tests
- [x] Update documentation

### CODE REVIEW

File: packages/core/src/components/card/card.tsx [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+ */
+
+import * as classNames from "classnames";
+import * as PureRender from "pure-render-decorator";
+import * as React from "react";
+import * as Classes from "../../common/classes";
+import { IProps } from "../../common/props";
+import { safeInvoke } from "../../common/utils";
+
+export type ElevationSize = 0 | 1 | 2 | 3 | 4;
```
Comment (JacekJagiello) [inline packages/core/src/components/card/card.tsx]: Should I introduce object witch contains these values? Something like `Intent` or `Classes`

File: packages/core/src/components/card/card.tsx [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+                return child;
+            }
+
+            return this.convertToElement(child);
+        });
+    }
+
+    private convertToElement(child: React.ReactChild): React.ReactElement<any> {
+        if (typeof child === "string" || typeof child === "number") {
+            return <p className={Classes.SKELETON}>{child}</p>;
```
Comment (JacekJagiello) [inline packages/core/src/components/card/card.tsx]: Have you other ideas for this case?

File: packages/core/test/card/cardTests.tsx [hunk 3]
```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ * Licensed under the BSD-3 License as modified (the “License”); you may obtain a copy
+ * of the license at https://github.com/palantir/blueprint/blob/master/LICENSE
+ * and https://github.com/palantir/blueprint/blob/master/PATENTS
+ */
+
+import { assert, expect } from "chai";
```
Comment (cmslewis) [inline packages/core/test/card/cardTests.tsx]: Use one or the other, not both.

File: packages/core/test/card/cardTests.tsx [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+    it("renders children", () => {
+        const expectedNumberOfChildren = 1;
+        const wrapper = shallow(
+            <Card>
+                <h4>Card content</h4>
+            </Card>,
+        );
+
+        const cardChildrens = wrapper.children().length;
```
Comment (cmslewis) [inline packages/core/test/card/cardTests.tsx]: Call this `numChildren` or get rid of the variable (it's only used once after all).

File: packages/core/test/card/cardTests.tsx [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        const wrapper = shallow(<Card elevation={3} interactive={true} className={Classes.TEXT_MUTED} />);
+
+        assert.isTrue(wrapper.hasClass(Classes.CARD), noClass(Classes.CARD));
+        assert.isTrue(wrapper.hasClass(Classes.ELEVATION_3), noClass(Classes.ELEVATION_3));
+        assert.isTrue(wrapper.hasClass(Classes.INTERACTIVE), noClass(Classes.INTERACTIVE));
+        assert.isTrue(wrapper.hasClass(Classes.TEXT_MUTED), noClass(Classes.TEXT_MUTED));
+    });
+
+    it("renders children", () => {
+        const expectedNumberOfChildren = 1;
```
Comment (cmslewis) [inline packages/core/test/card/cardTests.tsx]: don't need this IMO

File: packages/core/test/card/cardTests.tsx [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+                <h4>Card content</h4>
+            </Card>,
+        );
+
+        const cardChildrens = wrapper.children().length;
+
+        assert.equal(
+            cardChildrens,
+            expectedNumberOfChildren,
+            `Expected Card to have ${expectedNumberOfChildren} children, found ${cardChildrens} instead`,
```
Comment (cmslewis) [inline packages/core/test/card/cardTests.tsx]: If `assert.equal` fails, it'll already print an error like "expected 4 to equal 3", so we can shorten this all into one line:

```tsx
assert.equal(wrapper.children().length, 1, "rendered wrong number of children");
```

File: packages/core/test/card/cardTests.tsx [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        const errorMessage = "Exected onClick to be called, but it doesn't happend";
+        expect(onClick.calledOnce, errorMessage).to.be.true;
+    });
+
+    it("when loading, all childer becomes skeletons", () => {
+        const wrapper = shallow(
+            <Card loading={true}>
+                <h4>Sekelteon 1</h4>
+                <p>Sekelteon 2</p>
+                <span />
```
Comment (cmslewis) [inline packages/core/test/card/cardTests.tsx]: Ah cool, so the empty-element case is handled. What if you pass `{undefined}` as a child here?

File: packages/core/test/card/cardTests.tsx [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+        );
+
+        assert.equal(wrapper.children().length, 3);
+        assert.isTrue(wrapper.find("h4").hasClass(Classes.SKELETON), noClass(Classes.SKELETON));
+        assert.isTrue(wrapper.find("p").hasClass(Classes.SKELETON), noClass(Classes.SKELETON));
+        assert.isTrue(wrapper.find("p").hasClass(Classes.SKELETON), noClass(Classes.SKELETON));
+    });
+
+    it("when loading, children of type string or number are inserted into <p> elements", () => {
+        const cardWithString = shallow(<Card loading={true}>Sekelteon 1</Card>);
```
Comment (cmslewis) [inline packages/core/test/card/cardTests.tsx]: typo

File: packages/core/src/components/card/card.tsx (hunk 1 above)
Comment (giladgray) [reply packages/core/src/components/card/card.tsx]: I have no problem with the simple number union - it's the same thing I do for `iconSize` in `Icon`. Truth is that your `enum Elevation` above is equivalent and might still support the raw number usage.

File: packages/core/src/components/card/card.tsx [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+     * and the elevation shadow on the card increases by two levels.
+     * @default false
+     */
+    interactive?: boolean;
+
+    /**
+     * If set to `true`, all of the card's children will be converted into [skeletons](#core/components/progress/skeleton).
+     * @default false
+     */
+    loading?: boolean;
```
Comment (giladgray) [inline packages/core/src/components/card/card.tsx]: 👎 I'm incredibly wary of the implementation of this prop... Always found it brittle to modify `children` in any way. 

I'm inclined to nix this feature entirely and leave it up to consumers to do it themselves. Downside of this feature is that we could easily break their `children`, upside of not shipping it is that the consumer gets to ensure the correct behavior every time.

File: packages/core/src/components/card/card.tsx [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+        return React.Children.map(children, (child: React.ReactChild) => {
+            if (!loading) {
+                return child;
+            }
+
+            return this.convertToElement(child);
+        });
+    }
+
+    private convertToElement(child: React.ReactChild): React.ReactElement<any> {
```
[the remark this answers is not in the document]
Comment (giladgray) [reply packages/core/src/components/card/card.tsx]: I think I'd prefer to not do this at all. It's just too brittle.

File: packages/core/src/components/card/card.tsx (hunk 2 above)
Comment (giladgray) [reply packages/core/src/components/card/card.tsx]: I mean a `span` is infinitely better than `p` which has some associated block styles and margin.

File: packages/core/test/card/cardTests.tsx (hunk 6 above)
Comment (giladgray) [reply packages/core/test/card/cardTests.tsx]: better: `assert.lengthOf(wrapper.children(), 1)`. no need for message in this case.

File: packages/core/test/card/cardTests.tsx [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+ */
+
+import { assert, expect } from "chai";
+import { shallow } from "enzyme";
+import * as React from "react";
+
+import { Card, Classes } from "../../src/index";
+
+describe("<Card>", () => {
+    const noClass = (className: string) => `Card does not have "${className}" class`;
```
Comment (giladgray) [inline packages/core/test/card/cardTests.tsx]: this is nice but not necessary. I'd be fine with the assert message being just the `className`: we only need it to figure out _which_ assert failed.

`assert.isTrue(wrapper.hasClass(Classes.CARD), Classes.CARD)` will log `"pt-card"` on failure, which is enough info to debug.

File: packages/core/src/components/card/card.tsx [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+import * as React from "react";
+import * as Classes from "../../common/classes";
+import { IProps } from "../../common/props";
+import { safeInvoke } from "../../common/utils";
+
+export type ElevationSize = 0 | 1 | 2 | 3 | 4;
+
+export interface ICardProps extends IProps {
+    /**
+     * Elevation size of drop shadow that simulates card's height in the UI.
```
[the remark this answers is not in the document]
Comment (giladgray) [reply packages/core/src/components/card/card.tsx]: that's definitely an improvement. 
there might already be some words in the docs about card elevation?

File: packages/core/src/components/card/card.tsx (hunk 1 above)
Comment (llorca) [reply packages/core/src/components/card/card.tsx]: Let's keep the numbers way in case we update elevations levels. More descriptive key names is too restrictive

File: packages/core/src/components/card/card.tsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (llorca) [reply packages/core/src/components/card/card.tsx]: There's technically a `box-shadow` at elevation 0, so the last sentence could be confusing. Leave it out?

File: packages/core/src/components/card/card.tsx [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+     * @default 0
+     */
+    elevation?: ElevationSize;
+
+    /**
+     * Whether card on hover should change mouse to a pointer
+     * and the elevation shadow on the card increases by two levels.
```
[the remark this answers is not in the document]
Comment (llorca) [reply packages/core/src/components/card/card.tsx]: Remove `by two levels` to not be so restrictive on us

File: packages/core/src/components/card/card.tsx (hunk 2 above)
Comment (JacekJagiello) [reply packages/core/src/components/card/card.tsx]: Yeah, my first idea was to use `span`, but there was issue that `span` skeletons overlapped each other. But now, I'can't reproduce that again, maybe the problem was not correlated with `span` element itself.
So I'll switch `p` to `span` for now, but it need solid testing

File: packages/core/src/components/card/card.tsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (JacekJagiello) [reply packages/core/src/components/card/card.tsx]: I was inspired by reading [card css api](http://blueprintjs.com/docs/#core/components/card.css-api) ;)
I also like your description more

File: packages/core/src/components/card/card.tsx [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+            });
+        }
+
+        return this.cloneElement(child, {
+            className: classNames(child.props.className, Classes.SKELETON),
+        });
+    }
+
+    private cloneElement(element: React.ReactElement<any>, propsToOverwrite = {}) {
+        return React.cloneElement(element, { ...element.props, ...propsToOverwrite, tabIndex: "-1" });
```
[the remark this answers is not in the document]
Comment (JacekJagiello) [reply packages/core/src/components/card/card.tsx]: [skeletsons](http://blueprintjs.com/docs/#core/components/progress/skeleton) require `tabindex="-1"` to prevent from interacting with focusable elements

File: packages/core/test/card/cardTests.tsx (hunk 7 above)
Comment (JacekJagiello) [reply packages/core/test/card/cardTests.tsx]: Empty element are handled, but `{undefined}` ruins everything, good catch

File: packages/core/src/components/card/card.tsx (hunk 9 above)
Comment (JacekJagiello) [reply packages/core/src/components/card/card.tsx]: yeah, I agree that modifying children smells hacky. I will not mind to remove this feature, but first let's think about other ways we can do it.

1. Do you think better test coverage for multiple scenarios of nested elements can make brittle code more safe? If we write tests for most of use-cases, then at least brittle code is under our control.

2. Maybe there is another way to "replace" components with skeletons? I was thinking about doing kind of dump of card's inner structure and modify it instead, but that's another messy solution. We can also randomly generate skeletons, but that sucks from the UX perspective. Have you guys have other ideas?

3. Maybe loading can be done without using skeletons at all? In my project I've made this kind of loading for card:
spinner-loader

File: packages/core/src/components/card/card.tsx [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+            return this.convertToElement(child);
+        });
+    }
+
+    private convertToElement(child: React.ReactChild): React.ReactElement<any> {
+        if (typeof child === "string" || typeof child === "number") {
+            return <p className={Classes.SKELETON}>{child}</p>;
+        }
+
+        const childrenOfChild = child.props.children;
```
[the remark this answers is not in the document]
Comment (JacekJagiello) [reply packages/core/src/components/card/card.tsx]: Can be `grandchildren`, no problem ;)

File: packages/core/src/components/card/card.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (JacekJagiello) [reply packages/core/src/components/card/card.tsx]: I was also thinking about higher order component for card, that would do dirty job of replacing children of card with skeletons, but this still does not resolve the problem of brittle code. So that can of HoC shouldn't be added into library

File: packages/core/test/card/cardTests.tsx (hunk 11 above)
Comment (JacekJagiello) [reply packages/core/test/card/cardTests.tsx]: I will remove all exception messages from tests

File: packages/core/src/components/card/card.tsx (hunk 9 above)
Comment (giladgray) [reply packages/core/src/components/card/card.tsx]: That loading card transition is very cool! 

However I think that loading is simply out of scope for this `Card` component. It should be a lightweight wrapper around the CSS API and nothing more. 

A more general Skeleton solution should should be just that: more general. Definitely not baked into this one component, but available to be composed with any other BP component(s). This was the logic behind providing a single `.pt-skeleton` class instead of a full React component: slap the class on any element and _voila_ it's loading.

Comment (JacekJagiello) [reply packages/core/src/components/card/card.tsx]: I agree with you, let's keep the Card component simple. Consumers can always prepare two kinds of card's content, one with real data and one "skeletonized" and change them accordingly

Comment (giladgray) [reply packages/core/src/components/card/card.tsx]: exactly! glad we're in agreement.

File: packages/core/src/components/card/card.tsx [hunk 16]
```diff
@@ -57,57 +64,19 @@ export class Card extends React.Component<ICardProps, {}> {
...
-            className: classNames(child.props.className, Classes.SKELETON),
-        });
-    }
-
-    private cloneElement(element: React.ReactElement<any>, propsToOverwrite = {}) {
-        return React.cloneElement(element, { ...element.props, ...propsToOverwrite, tabIndex: "-1" });
-    }
-
-    private getClassName() {
+    private getClassName = () => {
```
Comment (giladgray) [inline packages/core/src/components/card/card.tsx]: this does not need to be bound. prefer `getClassName() {`.

File: packages/core/test/card/cardTests.tsx [hunk 17]
```diff
@@ -5,40 +5,31 @@
...
         const wrapper = shallow(<Card elevation={3} interactive={true} className={Classes.TEXT_MUTED} />);
 
-        assert.isTrue(wrapper.hasClass(Classes.CARD), noClass(Classes.CARD));
-        assert.isTrue(wrapper.hasClass(Classes.ELEVATION_3), noClass(Classes.ELEVATION_3));
-        assert.isTrue(wrapper.hasClass(Classes.INTERACTIVE), noClass(Classes.INTERACTIVE));
-        assert.isTrue(wrapper.hasClass(Classes.TEXT_MUTED), noClass(Classes.TEXT_MUTED));
+        assert.isTrue(wrapper.hasClass(Classes.CARD));
+        assert.isTrue(wrapper.hasClass(Classes.ELEVATION_3));
+        assert.isTrue(wrapper.hasClass(Classes.INTERACTIVE));
+        assert.isTrue(wrapper.hasClass(Classes.TEXT_MUTED));
```
Comment (giladgray) [inline packages/core/test/card/cardTests.tsx]: ok so the msg arg is useful here cuz the default message will be "expected false to be true."
add the second arg with the class name like I suggested a while back:
`assert.isTrue(wrapper.hasClass(Classes.TEXT_MUTED), Classes.TEXT_MUTED);`

File: packages/core/src/components/card/card.tsx [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+
+@PureRender
+export class Card extends React.Component<ICardProps, {}> {
+    public static displayName = "Blueprint.Card";
+
+    public static readonly ELEVATION_ZERO = 0 as 0;
+    public static readonly ELEVATION_ONE = 1 as 1;
+    public static readonly ELEVATION_TWO = 2 as 2;
+    public static readonly ELEVATION_THREE = 3 as 3;
+    public static readonly ELEVATION_FOUR = 4 as 4;
```
Comment (giladgray) [inline packages/core/src/components/card/card.tsx]: 🤔 not sure this is preferable to a single enum containing all possible values: `Elevation.ZERO`.

File: packages/core/src/components/card/card.tsx [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+        const { elevation, interactive } = this.props;
+        return classNames(
+            Classes.CARD,
+            {
+                [Classes.INTERACTIVE]: interactive,
+                [Classes.ELEVATION_0]: elevation === Card.ELEVATION_ZERO,
+                [Classes.ELEVATION_1]: elevation === Card.ELEVATION_ONE,
+                [Classes.ELEVATION_2]: elevation === Card.ELEVATION_TWO,
+                [Classes.ELEVATION_3]: elevation === Card.ELEVATION_THREE,
+                [Classes.ELEVATION_4]: elevation === Card.ELEVATION_FOUR,
```
Comment (giladgray) [inline packages/core/src/components/card/card.tsx]: maybe make a static array of elevation classes and then just index into it? `ELEVATION_CLASSES[elevation]`.

File: packages/core/test/card/cardTests.tsx [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+        assert.lengthOf(wrapper.children(), 1);
+        assert.equal(wrapper.children().html(), "<h4>Card content</h4>");
+    });
+
+    it("call onClick when card is clicked", () => {
+        const onClick = sinon.spy();
+        const wrapper = shallow(<Card onClick={onClick} />);
+
+        wrapper.simulate("click");
+
```
Comment (giladgray) [inline packages/core/test/card/cardTests.tsx]: remove all this whitespace.

in fact you could just write `shallow(<Card onClick={onClick} />).simulate("click");`

File: packages/core/src/components/card/card.md [hunk 21]
```diff
@@ -21,3 +21,25 @@ the card increases by two levels.
...
+Make sure to review the [general usage docs for JS components](#blueprint.usage).
+
+This component is a simple wrapper around the CSS API that abstracts away the HTML complexity.
+
+```
+<Card interactive={true} elevation={Card.ELEVATION_TWO}>
+    <h5><a href="#">Card heading</a></h5>
+    <p>
+    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Quisque eget tortor felis.
+    Fusce dapibus metus in dapibus mollis. Quisque eget ex diam.
```
Comment (giladgray) [inline packages/core/src/components/card/card.md]: indent these lines another 4 spaces cuz they're inside the `p` tag

File: packages/core/src/components/card/card.tsx [hunk 22]
```diff
@@ -0,0 +1,88 @@
...
+    interactive?: boolean;
+
+    /**
+     * Callback invoked when the card is clicked.
+     * Recommended when `interactive` is `true`.
+     */
+    onClick?: (e: React.MouseEvent<HTMLDivElement>) => void;
+}
+
+/* tslint:disable:object-literal-sort-keys */
```
Comment (JacekJagiello) [inline packages/core/src/components/card/card.tsx]: Good? Bad? :)

File: packages/core/examples/cardExample.tsx [hunk 23]
```diff
@@ -65,13 +65,9 @@ export class CardExample extends BaseExample<ICardExampleState> {
...
-        this.updateState({ elevation });
-    };
+    private handleElevationChange = (elevation: number) => this.updateState({ elevation });
 
-    private handleInteractiveChange = () => {
-        this.updateState({ interactive: !this.state.interactive });
-    };
+    private handleInteractiveChange = () => this.updateState({ interactive: !this.state.interactive });
 
     private updateState(state: Partial<ICardExampleState>) {
```
Comment (giladgray) [inline packages/core/examples/cardExample.tsx]: this is not necessary... React essentially does this internally. why did you add this?

File: packages/core/src/components/card/card.tsx (hunk 22 above)
Comment (giladgray) [reply packages/core/src/components/card/card.tsx]: bad. `export enum Elevation` so it's an actual type.

File: packages/core/examples/cardExample.tsx [hunk 24]
```diff
@@ -0,0 +1,75 @@
...
+
+import * as React from "react";
+
+import { Card, Classes, ElevetaionSize, Slider, Switch } from "@blueprintjs/core";
+import { BaseExample } from "@blueprintjs/docs";
+
+export interface ICardExampleState {
+    elevation: ElevetaionSize;
+    interactive: boolean;
+    onClick: (e: React.MouseEvent<HTMLElement>) => void;
```
Comment (giladgray) [inline packages/core/examples/cardExample.tsx]: make all state fields optional `elevation?` (we use old React typings)

File: packages/core/examples/cardExample.tsx [hunk 25]
```diff
@@ -0,0 +1,75 @@
...
+        ];
+    }
+
+    private handleElevationChange = (elevation: number) => this.updateState({ elevation });
+
+    private handleInteractiveChange = () => this.updateState({ interactive: !this.state.interactive });
+
+    private updateState(state: Partial<ICardExampleState>) {
+        return this.setState({ ...this.state, ...state });
+    }
```
Comment (giladgray) [inline packages/core/examples/cardExample.tsx]: remove this function after making all state fields optional.

File: packages/core/src/components/card/card.md [hunk 26]
```diff
@@ -21,3 +21,25 @@ the card increases by two levels.
 Users expect an interactive card to be a single clickable unit.
 
 @css pt-card.pt-interactive
+
+@## JavaScript API
+
+Then `Card` component is available in the __@blueprintjs/core__ package.
+Make sure to review the [general usage docs for JS components](#blueprint.usage).
+
+This component is a simple wrapper around the CSS API that abstracts away the HTML complexity.
```
Comment (giladgray) [inline packages/core/src/components/card/card.md]: drop the last six words, the HTML is not complex. "around the CSS API."

File: packages/core/src/components/card/card.md [hunk 27]
```diff
@@ -21,3 +21,25 @@ the card increases by two levels.
...
+
+```
+<Card interactive={true} elevation={Card.ELEVATION_TWO}>
+    <h5><a href="#">Card heading</a></h5>
+    <p>
+        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Quisque eget tortor felis.
+        Fusce dapibus metus in dapibus mollis. Quisque eget ex diam.
+    </p>
+    <button type="button" className={Classes.BUTTON}>Submit</button>
```
Comment (giladgray) [inline packages/core/src/components/card/card.md]: there's a component for this: `<Button>Submit</Button>`

File: packages/core/src/components/card/card.md [hunk 28]
```diff
@@ -21,3 +21,25 @@ the card increases by two levels.
...
+
+This component is a simple wrapper around the CSS API that abstracts away the HTML complexity.
+
+```
+<Card interactive={true} elevation={Card.ELEVATION_TWO}>
+    <h5><a href="#">Card heading</a></h5>
+    <p>
+        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Quisque eget tortor felis.
+        Fusce dapibus metus in dapibus mollis. Quisque eget ex diam.
+    </p>
```
Comment (giladgray) [inline packages/core/src/components/card/card.md]: let's keep it simple and just say `<p>Card content</p>`.

File: packages/core/src/components/card/card.tsx [hunk 29]
```diff
@@ -0,0 +1,88 @@
...
+ */
+
+import * as classNames from "classnames";
+import * as PureRender from "pure-render-decorator";
+import * as React from "react";
+import * as Classes from "../../common/classes";
+import { IProps } from "../../common/props";
+import { safeInvoke } from "../../common/utils";
+
+export type Elevation = 0 | 1 | 2 | 3 | 4;
```
Comment (giladgray) [inline packages/core/src/components/card/card.tsx]: delete this, replace with `enum Elevation`.

File: packages/core/src/components/card/card.tsx [hunk 30]
```diff
@@ -0,0 +1,88 @@
...
+    Classes.ELEVATION_4,
+];
+
+@PureRender
+export class Card extends React.Component<ICardProps, {}> {
+    public static displayName = "Blueprint.Card";
+
+    public render() {
+        return (
+            <div className={this.getClassName()} onClick={this.handleClick}>
```
Comment (giladgray) [inline packages/core/src/components/card/card.tsx]: `onClick={this.props.onClick}` and remove `this.handleClick`. React "safe invokes" internally.

File: packages/core/src/components/card/card.tsx [hunk 31]
```diff
@@ -0,0 +1,88 @@
...
+            </div>
+        );
+    }
+
+    private handleClick = (e: React.MouseEvent<HTMLElement>) => {
+        safeInvoke(this.props.onClick, e);
+    };
+
+    private getClassName() {
+        const { elevation, interactive } = this.props;
```
Comment (giladgray) [inline packages/core/src/components/card/card.tsx]: include `className`

File: packages/core/src/components/card/card.tsx (hunk 19 above)
Comment (llorca) [reply packages/core/src/components/card/card.tsx]:
```js
Classes[`ELEVATION_${elevation}`]
```
? but then we lose the ability to search for Classes.X in the codebase

**Comment by JacekJagiello** [comment]
@giladgray @cmslewis Any more thoughts?

File: packages/core/src/components/card/card.md (hunk 28 above)
Comment (giladgray) [reply packages/core/src/components/card/card.md]: can you put this all on one line?

File: packages/core/examples/cardExample.tsx [hunk 32]
```diff
@@ -0,0 +1,71 @@
...
+ * and https://github.com/palantir/blueprint/blob/master/PATENTS
+ */
+
+import * as React from "react";
+
+import { Card, Classes, Elevetaion, Slider, Switch } from "@blueprintjs/core";
+import { BaseExample } from "@blueprintjs/docs";
+
+export interface ICardExampleState {
+    elevation?: Elevetaion;
```
Comment (giladgray) [inline packages/core/examples/cardExample.tsx]: spelling: `Elevation`

this doesn't actually compile, but webpack errors don't fail the build 😢 
```
ERROR in ./packages/core/examples/cardExample.tsx
(10,25): error TS2305: Module '"/home/ubuntu/blueprint/packages/site-docs/node_modules/@blueprintjs/core/dist/index"' has no exported member 'Elevetaion'.
```

File: packages/core/src/components/card/card.md [hunk 33]
```diff
@@ -21,3 +21,24 @@ the card increases by two levels.
...
+Make sure to review the [general usage docs for JS components](#blueprint.usage).
+
+This component is a simple wrapper around the CSS API.
+
+```
+<Card interactive={true} elevation={Card.ELEVATION_TWO}>
+    <h5><a href="#">Card heading</a></h5>
+    <p>
+        Card content
+    </p>
```
Comment (giladgray) [inline packages/core/src/components/card/card.md]: one line instead of three. `<p>Card Content</p>`

**Comment by JacekJagiello** [comment]
Thanks for review @giladgray :) @cmslewis can you also review changes? No rush

File: packages/core/src/components/card/card.tsx [hunk 34]
```diff
@@ -0,0 +1,79 @@
...
+    Classes.ELEVATION_2,
+    Classes.ELEVATION_3,
+    Classes.ELEVATION_4,
+];
+
+@PureRender
+export class Card extends React.Component<ICardProps, {}> {
+    public static displayName = "Blueprint.Card";
+
+    public render() {
```
Comment (cmslewis) [inline packages/core/src/components/card/card.tsx]: Be sure to define your `defaultProps` for `elevation` and `interactive`.

File: packages/core/src/components/card/card.tsx [hunk 35]
```diff
@@ -0,0 +1,79 @@
...
+            </div>
+        );
+    }
+
+    private getClassName() {
+        const { elevation, interactive, className } = this.props;
+        return classNames(
+            Classes.CARD,
+            { [Classes.INTERACTIVE]: interactive },
+            ELEVATION_CLASSES[elevation],
```
Comment (cmslewis) [inline packages/core/src/components/card/card.tsx]: This would return `undefined` if you don't set `elevation` to `0` by default.

File: packages/core/test/card/cardTests.tsx [hunk 36]
```diff
@@ -0,0 +1,40 @@
...
+            <Card>
+                <h4>Card content</h4>
+            </Card>,
+        );
+
+        assert.lengthOf(wrapper.children(), 1);
+        assert.equal(wrapper.children().html(), "<h4>Card content</h4>");
+    });
+
+    it("call onClick when card is clicked", () => {
```
Comment (cmslewis) [inline packages/core/test/card/cardTests.tsx]: Change `call` to `calls`.

File: packages/core/src/components/card/card.tsx [hunk 37]
```diff
@@ -0,0 +1,79 @@
...
+     * the elevation, the higher the drop shadow. At elevation `0`, no drop
+     * shadow is applied.
+     * 
+     * @default 0
+     */
+    elevation?: Elevation;
+
+    /**
+     * Whether the card should respond to user interactions. If set to `true`,
+     * hovering over the card will increase the card's elevation by two levels
```
Comment (cmslewis) [inline packages/core/src/components/card/card.tsx]: Oh, let's also remove `by two levels` per @llorca's request.

File: packages/core/examples/cardExample.tsx [hunk 38]
```diff
@@ -0,0 +1,68 @@
...
+    };
+
+    protected renderExample() {
+        return (
+            <Card {...this.state}>
+                <h5>
+                    <a href="#">Card heading</a>
+                </h5>
+                <p>Lorem ipsum dolor sit amet, consectetur adipiscing elit.</p>
+                <button type="button" className={Classes.BUTTON}>
```
Comment (llorca) [inline packages/core/examples/cardExample.tsx]: `<Button>`

File: packages/core/src/components/card/card.tsx [hunk 39]
```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ * Licensed under the BSD-3 License as modified (the “License”); you may obtain a copy
+ * of the license at https://github.com/palantir/blueprint/blob/master/LICENSE
+ * and https://github.com/palantir/blueprint/blob/master/PATENTS
```
Comment (giladgray) [inline packages/core/src/components/card/card.tsx]: semantic merge conflict: we just relicensed to Apache 2.0 🙊

File: packages/core/src/components/card/card.tsx (hunk 34 above)
Comment (giladgray) [reply packages/core/src/components/card/card.tsx]: yes default for `elevation` is critical

File: packages/core/test/card/cardTests.tsx [hunk 40]
```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ * Licensed under the BSD-3 License as modified (the “License”); you may obtain a copy
+ * of the license at https://github.com/palantir/blueprint/blob/master/LICENSE
+ * and https://github.com/palantir/blueprint/blob/master/PATENTS
```
Comment (giladgray) [inline packages/core/test/card/cardTests.tsx]: relicense

**Comment by giladgray** [comment]
@JacekJagiello comments above feel like the last round of feedback. would you kindly resolve them and we'll get this bad boy merged? 🍻

**Comment by JacekJagiello** [comment]
@giladgray @cmslewis It's done ;)

**Comment by JacekJagiello** [comment]
@cmslewis Oh I've missed that, fixing now, will be ready in the minute :)

**Comment by cmslewis** [comment]
@giladgray  take a look when you're ready

**Comment by cmslewis** [comment]
Thanks @JacekJagiello!

File: packages/core/src/components/card/card.tsx [hunk 41]
```diff
@@ -0,0 +1,82 @@
...
+    interactive?: boolean;
+
+    /**
+     * Callback invoked when the card is clicked.
+     * Recommended when `interactive` is `true`.
+     */
+    onClick?: (e: React.MouseEvent<HTMLDivElement>) => void;
+}
+
+export enum Elevation {
```
Comment (adidahiya) [inline packages/core/src/components/card/card.tsx]: this enum feels pretty excessive. I think this should've been a numeric literal union type, `type Elevation  = 0 | 1 | 2 | 3 | 4`. we should make this change in 2.0

File: packages/core/test/card/cardTests.tsx [hunk 42]
```diff
@@ -0,0 +1,39 @@
...
+ * Licensed under the terms of the LICENSE file distributed with this project.
+ */
+
+import { assert } from "chai";
+import { shallow } from "enzyme";
+import * as React from "react";
+
+import { Card, Classes } from "../../src/index";
+
+describe("<Card>", () => {
```
Comment (adidahiya) [inline packages/core/test/card/cardTests.tsx]: These are not very useful tests and IMO are a good example of how unit test coverage thresholds can be arbitrary. The maintenance + performance cost of these tests isn't worth what they provide us. Let's remove them + use `/* istanbul ignore next */` directives and/or lower the coverage thresholds.
